### PR TITLE
[CDSID - 535] - Tests added to all GETS of the Goods Item page

### PR DESCRIPTION
**Pull request by raghu1978**

### CODE REVIEW

File: app/views/goods_items_origins.scala.html [hunk 1]
```diff
@@ -30,7 +30,7 @@
 
 <h2 class="heading-medium">Government Agency Goods Item Origin of goods	</h2>
 @if(origins.nonEmpty) {
-No of Goods Item origins added : <td scope="row">@origins.size</td>
+No of Goods Item Origins added : <td scope="row">@origins.size</td>
```
Comment (kurtlogan) [inline app/views/goods_items_origins.scala.html]: `td` should not be placed outside of a table

File: app/views/goods_items_packagings.scala.html [hunk 2]
```diff
@@ -30,7 +30,7 @@
 
 <h2 class="heading-medium">Government Agency Goods Item Packaging	</h2>
 @if(packagings.nonEmpty) {
-No of Goods Item Packaging added : <td scope="row">@packagings.size</td>
+No of Goods Item Packagings added : <td scope="row">@packagings.size</td>
```
Comment (kurtlogan) [inline app/views/goods_items_packagings.scala.html]: `td` should not be placed outside of a table

File: app/views/gov_agency_goods_items_add_docs.scala.html [hunk 3]
```diff
@@ -30,7 +30,7 @@
 
 <h2 class="heading-medium">Government Agency Goods Item Additional Documents</h2>
 @if(additionalDocuments.nonEmpty) {
-No of Goods Item additional docs added : <td scope="row">@additionalDocuments.size</td>
+No of Goods Item Additional Documents added : <td scope="row">@additionalDocuments.size</td>
```
Comment (kurtlogan) [inline app/views/gov_agency_goods_items_add_docs.scala.html]: Again `td`

File: test/controllers/GovernmentAgencyGoodsItemsControllerSpec.scala [hunk 4]
```diff
@@ -170,10 +175,10 @@ class GovernmentAgencyGoodsItemsControllerSpec extends CustomsSpec
     "display good items page with pre-populated data on revisiting the page " in withFeatures(enabled(Feature.submit)) {
       withSignedInUser() { (headers, session, tags) =>
         withCaching(Some(goodsItemValueInformation))
-        withRequest("GET", goodsItemsPageUri, headers, session, tags) { resp =>
-         //  includesHtmlInput(resp, "customsValueAmount",value = "30.00")
-           includesHtmlInput(resp, "sequenceNumeric",value = "123")
-           includesHtmlInput(resp, "statisticalValueAmount.currencyId",value = "GBP")
+        withRequest(get, goodsItemsPageUri, headers, session, tags) { resp =>
+          //  includesHtmlInput(resp, "customsValueAmount",value = "30.00")
```
Comment (kurtlogan) [inline test/controllers/GovernmentAgencyGoodsItemsControllerSpec.scala]: Remove commented out code

File: test/controllers/GovernmentAgencyGoodsItemsControllerSpec.scala [hunk 5]
```diff
@@ -226,18 +231,258 @@ class GovernmentAgencyGoodsItemsControllerSpec extends CustomsSpec
       "/customs-declare-imports/submit-declaration-goods/gov-agency-goods-items")
     assertNavigation(Map("add" -> "wrong-url"), "", Status.BAD_REQUEST)
   }
-  
 
-  private def assertNavigation(payload :Map[String,String] ,headerLocationUri:String,
-    respStatus :Int = Status.SEE_OTHER, postUri :String = navigateToSelectedGoodsItemPageUri) =
+  "show gov-agency-goods-items-additional-docs fields on navigating to the screen" in withFeatures((enabled(Feature.submit))) {
+    withSignedInUser() { (headers, session, tags) =>
```
Comment (kurtlogan) [inline test/controllers/GovernmentAgencyGoodsItemsControllerSpec.scala]: There doesn't appear to be much variation in what is being tested here. it looks like only the happy path has been covered and most of the assertions are against html inputs being rendered to the page. It is also quite a difficult file to follow. Consider perhaps testing the view separately, including tests for when data is loaded from the cache and for when users are not logged or are missing things like enrolments or identifiers

File: test/controllers/GovernmentAgencyGoodsItemsControllerSpec.scala [hunk 6]
```diff
@@ -226,18 +231,258 @@ class GovernmentAgencyGoodsItemsControllerSpec extends CustomsSpec
...
+      }
+    }
+  }
+
+  "pre-populate gov-agency-goods-items-additional-docs as a list on navigating to the screen" in withFeatures((enabled(Feature.submit))) {
+    withSignedInUser() { (headers, session, tags) =>
+      withCaching(Some(goodsItem))
+      withRequest(get, goodsItemsAdditionalDocsPageUri, headers, session, tags) { resp =>
+        val content = contentAsHtml(resp)
+        contentAsString(resp) must include("No of Goods Item Additional Documents added : <td scope=\"row\">1")
```
Comment (kurtlogan) [inline test/controllers/GovernmentAgencyGoodsItemsControllerSpec.scala]: You have tested here that the number 1 is displayed when a single item has been entered but not any of the content rendered. Consider adding more tests like that

This is the same for other tests in this file

File: test/controllers/TestData.scala [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+ * limitations under the License.
+ */
+
+package controllers
+
+import domain.{GoodsItemValueInformation, GovernmentAgencyGoodsItem}
+import org.joda.time.DateTime
+import uk.gov.hmrc.wco.dec._
+
+trait TestData {
```
Comment (kurtlogan) [inline test/controllers/TestData.scala]: Personally I dislike this style of testing with fixed data. We have imported scalacheck for use on our ticket so we can make use of property based testing. Consider changing this in future to use generators and property based tests

File: app/views/goods_items_origins.scala.html (hunk 1 above)
Comment (raghu1978) [reply app/views/goods_items_origins.scala.html]: I am planning to pass the frontend stuff to Russel but will change per comment

File: app/views/goods_items_packagings.scala.html (hunk 2 above)
Comment (raghu1978) [reply app/views/goods_items_packagings.scala.html]: same as above

File: test/controllers/GovernmentAgencyGoodsItemsControllerSpec.scala (hunk 5 above)
Comment (raghu1978) [reply test/controllers/GovernmentAgencyGoodsItemsControllerSpec.scala]: I have planned to move the gets and posts separately that's one reason I left all the get and view related in this file

File: test/controllers/GovernmentAgencyGoodsItemsControllerSpec.scala (hunk 6 above)
Comment (raghu1978) [reply test/controllers/GovernmentAgencyGoodsItemsControllerSpec.scala]: that's is something I am going refactor /include as I currently don't know how I can use JSOUP to cover those tests

File: test/controllers/TestData.scala (hunk 7 above)
Comment (raghu1978) [reply test/controllers/TestData.scala]: I will have a look and update cheers

**Comment by raghu1978** [comment]
> From what I could see testing didn't quite cover the kinds of scenarios I would have expected it to. Happy to discuss what I mean with you if my comments are not clear

I didnt finish the testing the PR i have is to Just cover basic test cases for views and rendering, i am on the same view that i need different files as the tests would be too many in one file and difficult to follow.

File: app/views/goods_items_origins.scala.html (hunk 1 above)
Comment (raghu1978) [reply app/views/goods_items_origins.scala.html]: fixed and updated code/tests  accordingly

**Pull request merged**